File: android/app/src/main/java/co/edgesecure/app/EdgeCore.java
package co.edgesecure.app;

import android.content.Context;
import android.util.Base64;
import android.util.Log;
import androidx.annotation.NonNull;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class EdgeCore {
  Context context;
  String logId;

  public EdgeCore(String logId, Context context) {
    this.context = context;
    this.logId = logId;
  }

  /**
   * Fetches messages from the auth server, and returns an array of usernames with pending logins.
   */
  public @NonNull Iterable<String> fetchPendingLogins() throws JSONException, IOException {
    File basePath = context.getFilesDir();
    File loginsPath = new File(basePath, "logins");
    File[] files = loginsPath.listFiles();

    // Load the files:
    HashMap<String, String> loginIds = new HashMap();
    for (File file : files) {
      try {
        JSONObject json = new JSONObject(readFile(file));
        if (!json.has("loginAuthBox")) continue;
        String loginId = json.getString("loginId");
        String username = json.getString("username");
        loginIds.put(loginId, username);
      } catch (Exception e) {
        continue;
      }
    }

    // Bail out if there are no logged-in users:
    if (loginIds.size() == 0) return new ArrayList();

    // Prepare our payload:
    JSONObject payloadJson = new JSONObject();
    payloadJson.put("loginIds", new JSONArray(loginIds.keySet()));

    // Do the request:
    String uri = "https://auth.airbitz.co/api/v2/messages";
    JSONObject replyJson = new JSONObject(authFetch(uri, payloadJson.toString()));

    // Validate the response:
    int statusCode = replyJson.getInt("status_code");
    if (statusCode != 0) throw new JSONException("Incorrect status code");
    JSONArray messages = replyJson.getJSONArray("results");

    // Find messages with problems:
    ArrayList<String> problemUsers = new ArrayList();
    int messagesLength = messages.length();
    for (int i = 0; i < messagesLength; ++i) {
      JSONObject message = messages.getJSONObject(i);
      String loginId = message.getString("loginId");
      String username = loginIds.get(loginId);
      if (username == null) continue;

      JSONArray pendingVouchers = message.optJSONArray("pendingVouchers");
      boolean hasVoucher = pendingVouchers != null && pendingVouchers.length() > 0;
      boolean hasReset = message.optBoolean("otpResetPending", false);

      if (hasVoucher || hasReset) {
        problemUsers.add(username);
      }
    }

    return problemUsers;
  }

  public void updatePushToken(String token) throws JSONException, IOException {
    // Read the clientId:
    File basePath = context.getFilesDir();
    File file = new File(basePath, "client.json");
    JSONObject json = new JSONObject(readFile(file));
    String clientId64 = json.getString("clientId");
    String clientId = Base58.encode(Base64.decode(clientId64, Base64.DEFAULT));

    // Prepare our payload:
    JSONObject payloadJson = new JSONObject();
    payloadJson.put("apiKey", EdgeApiKey.apiKey);
    payloadJson.put("deviceId", clientId);
    payloadJson.put("deviceToken", token);

    // Do the request:
    String uri = EdgeApiKey.pushServer.concat("/v2/device/");
    pushFetch(uri, payloadJson.toString());
  }

  /** Does a request / reply with the auth server. */
  private String authFetch(String uri, String body) throws IOException {
    HttpsURLConnection connection = null;
    try {
      // Set up the HTTPS connection:
      connection = (HttpsURLConnection) new URL(uri).openConnection();
      SSLContext context = SSLContext.getInstance("TLS");
      context.init(null, null, null);
      connection.setSSLSocketFactory(context.getSocketFactory());

      // Add the auth server headers:
      byte[] bodyData = body.getBytes("UTF-8");
      connection.setRequestProperty("Accept", "application/json");
      connection.setRequestProperty("Content-Type", "application/json");
      connection.setRequestProperty("Authorization", "Token " + EdgeApiKey.apiKey);
      connection.setRequestProperty("Content-Length", Integer.toString(bodyData.length));
      connection.setRequestMethod("POST");
      connection.setDoInput(true);
      connection.setDoOutput(true);
      connection.setUseCaches(false);

      // Send the body:
      OutputStream wr = connection.getOutputStream();
      wr.write(bodyData);
      wr.flush();
      wr.close();
      connection.connect();

      Log.i(logId, String.format("%s %d", uri, connection.getResponseCode()));

      // Read the reply body:
      return readInputStream(connection.getInputStream());
    } catch (Exception e) {
      throw new IOException("Could not reach auth server " + uri, e);
    } finally {
      if (connection != null) connection.disconnect();
    }
  }

  /** Does a request / reply with the push server. */
  private void pushFetch(String uri, String body) throws IOException {
    HttpsURLConnection connection = null;
    try {
      // Set up the HTTPS connection:
      connection = (HttpsURLConnection) new URL(uri).openConnection();
      SSLContext context = SSLContext.getInstance("TLS");
      context.init(null, null, null);
      connection.setSSLSocketFactory(context.getSocketFactory());

      // Add the auth server headers:
      byte[] bodyData = body.getBytes("UTF-8");
      connection.setRequestProperty("Accept", "application/json");
      connection.setRequestProperty("Content-Type", "application/json");
      connection.setRequestProperty("Content-Length", Integer.toString(bodyData.length));
      connection.setRequestMethod("POST");
      connection.setDoInput(true);
      connection.setDoOutput(true);
      connection.setUseCaches(false);

      // Send the body:
      OutputStream wr = connection.getOutputStream();
      wr.write(bodyData);
      wr.flush();
      wr.close();
      connection.connect();

      Log.i(logId, String.format("%s %d", uri, connection.getResponseCode()));
    } catch (Exception e) {
      throw new IOException("Could not reach push server " + uri, e);
    } finally {
      if (connection != null) connection.disconnect();
    }
  }

  /** Reads a file from disk. */
  private static @NonNull String readFile(File file) throws IOException {
    try (FileInputStream stream = new FileInputStream(file)) {
      return readInputStream(stream);
    }
  }

  /** Reads an input stream into a string, as utf-8. */
  private static @NonNull String readInputStream(@NonNull InputStream stream) throws IOException {
    ByteArrayOutputStream out = new ByteArrayOutputStream();

    int size;
    byte[] data = new byte[4096];
    while ((size = stream.read(data)) > 0) {
      out.write(data, 0, size);
    }

    return out.toString("UTF-8");
  }
}

File: ios/EdgeCore.swift
import Foundation

@objc
public class EdgeCore: NSObject {
  @objc
  public func fetchPendingLogins(completion: @escaping ([NSString]?) -> Void) {
    do {
      let decoder = JSONDecoder()
      let encoder = JSONEncoder()

      // List the logins folder:
      let fs = FileManager.default
      guard let documentUrl = fs.urls(for: .documentDirectory, in: .userDomainMask).first else {
        return completion(nil)
      }
      let loginsUrl = documentUrl.appendingPathComponent("logins").absoluteURL
      let files = try fs.contentsOfDirectory(
        at: loginsUrl, includingPropertiesForKeys: nil, options: [])

      // Load the files:
      var loginIds: [String: String] = [:]
      for file in files {
        if let data = try? Data(contentsOf: file),
          let loginStash = try? decoder.decode(LoginStash.self, from: data)
        {
          if loginStash.loginAuthBox != nil {
            loginIds[loginStash.loginId] = loginStash.username
          }
        }
      }

      // Bail out if there are no logged-in users:
      if loginIds.count == 0 { return completion([]) }

      // Prepare our payload:
      var body = LoginRequestBody()
      body.loginIds = Array.init(loginIds.keys)

      // Prepare our request:
      guard let url = URL(string: "https://login.edge.app/api/v2/messages") else {
        return completion(nil)
      }
      var request = URLRequest(url: url)
      request.httpBody = try encoder.encode(body)
      request.httpMethod = "POST"
      request.setValue("application/json", forHTTPHeaderField: "Accept")
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
      request.setValue("Token \(EdgeApiKey.apiKey)", forHTTPHeaderField: "Authorization")

      // Perform the request:
      let task = URLSession.shared.dataTask(with: request) { data, response, error in
        if error != nil { return completion(nil) }
        if let data = data,
          let body = try? decoder.decode(LoginResponseBody.self, from: data),
          let results = body.results
        {
          var usernames: [NSString] = []
          for result in results {
            if result.otpResetPending == true || result.pendingVouchers?.count ?? 0 > 0 {
              if let username = loginIds[result.loginId] {
                usernames.append(NSString(string: username))
              }
            }
          }
          completion(usernames)
        } else {
          completion(nil)
        }
      }

      task.resume()
    } catch {
      return completion(nil)
    }
  }

  @objc
  public func updatePushToken(token: NSString, completion: @escaping (Bool) -> Void) {
    do {
      let encoder = JSONEncoder()

      // Read the clientId:
      guard let clientId = readClientId() else {
        return completion(false)
      }

      // Prepare our payload:
      var body = PushRequestBody()
      body.apiKey = EdgeApiKey.apiKey
      body.deviceId = Base58.encode(Array(clientId))
      body.deviceToken = String(token)

      // Prepare our request:
      guard let url = URL(string: "\(EdgeApiKey.pushServer)/v2/device/") else {
        return completion(false)
      }
      var request = URLRequest(url: url)
      request.httpBody = try encoder.encode(body)
      request.httpMethod = "POST"
      request.setValue("application/json", forHTTPHeaderField: "Accept")
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")

      // Perform the request:
      let task = URLSession.shared.dataTask(with: request) { data, response, error in
        if error != nil { return completion(false) }
        if let httpResponse = response as? HTTPURLResponse {
          if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            return completion(false)
          }
        }
        return completion(true)
      }

      task.resume()
    } catch {
      return completion(false)
    }
  }

  private func readClientId() -> Data? {
    let decoder = JSONDecoder()
    let fs = FileManager.default
    guard let documentUrl = fs.urls(for: .documentDirectory, in: .userDomainMask).first else {
      return nil
    }
    let clientUrl = documentUrl.appendingPathComponent("client.json").absoluteURL
    if let data = try? Data(contentsOf: clientUrl),
      let clientFile = try? decoder.decode(ClientFile.self, from: data),
      let clientId = Data(base64Encoded: clientFile.clientId)
    {
      return clientId
    }
    return nil
  }

  private struct ClientFile: Codable {
    var clientId: String
  }

  private struct EdgeBox: Codable {
    // TODO
  }

  private struct EdgePendingVoucher: Codable {
    // TODO
  }

  private struct LoginRequestBody: Codable {
    var loginIds: [String]?
  }

  private struct LoginResponseBody: Codable {
    var results: [MessagesPayload]?
    var status_code: Int
    var message: String
  }

  private struct LoginStash: Codable {
    var loginId: String
    var username: String
    var loginAuthBox: EdgeBox?
  }

  private struct MessagesPayload: Codable {
    var loginId: String
    var otpResetPending: Bool?
    var pendingVouchers: [EdgePendingVoucher]?
    var recovery2Corrupt: Bool?
  }

  private struct PushRequestBody: Codable {
    var apiKey: String?
    var deviceId: String?
    var deviceToken: String?
  }
}
